### PR TITLE
example app stateful-sidebar-with-dialog

### PR DESCRIPTION
**Pull request by Muzietto**
## Purpose

A very simple example Contentful app that demonstrates a way to share (indirectly!) a React state between two component locations.

## Approach

It is not possible to create a state+dispatcher with useReducer and then truly share them between two app locations in the way two "normal" React components would do.

This solution passes a serialized (JSON.stringify) copy of the state from the Sidebar to the Dialog, which uses JSON.parse to recover it; the state is serialized again and sent back to the Sidebar when the Dialog closes.

## Testing steps

This code is self-containing and it does not interact with other code from this repository in any manner.

This example app has been created with "create-contentful-app"; it must be prepared with "npm install" and runs with "react start". You need of course to create a localhost app definition in the Contentful webapp.

## Breaking Changes

None.

## Dependencies and/or References

This is the result of ticket https://support.contentful.com/hc/en-us/requests/174031

## Deployment

This example app has been created with "create-contentful-app"; it must be prepared with "npm install" and runs with "react start". You need of course to create a localhost app definition in the Contentful webapp.

### CODE REVIEW

**Comment by Muzietto** [comment]
Sorry, but I don't understand how I could improve my code to let it pass the two failing test.

Please give me some hint.

**Comment by david-shibley-contentful** [comment]
@Muzietto looking into this now 👀

**Comment by david-shibley-contentful** [comment]
@Muzietto the reason the tests were failing was because you are an outside contributor so the circle ci job didn't have access to restricted context. I was able to resolve that issue by running the job manually. Getting past the tests there is still a failure due to prettier. Can you please format the following files:

- index.jsx
- Dialog.jsx
- Sidebar.jsx
- reducerAndInitialSidebarState.js

**Comment by Muzietto** [comment]
This is the .prettierrc I am using, straight from your repository: 

{
  "printWidth": 100,
  "tabWidth": 2,
  "useTabs": false,
  "singleQuote": true,
  "bracketSameLine": true,
  "trailingComma": "es5"
}

I have formatted manually again those files explicitly with Prettier, but this hasn't brought changes.

What is the .prettierrc used by your CI ?

**Comment by david-shibley-contentful** [comment]
Strange. That is the correct prettier config 🤔

File: examples/stateful-sidebar-with-dialog/src/index.jsx [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+const root = createRoot(container);
+
+if (process.env.NODE_ENV === 'development' && window.self === window.top) {
+  // You can remove this if block before deploying your app
+  root.render(<LocalhostWarning />);
+} else {
+  root.render(
+    <SDKProvider>
+      <GlobalStyles />
+        <App />
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/index.jsx:21]:
```suggestion
      <GlobalStyles />
      <App />
```

File: examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+      dispatch({
+        type: 'SET_TIME',
+        payload: new Date().toString(),
+      });
+    }}>SET TIME</Button>
+    <Button
+      onClick={() => {
+        sdk.close(JSON.stringify(state));
+      }}>CLOSE</Button>
+  </>;
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx:36]:
```suggestion
    <Button
      onClick={() => {
        sdk.close(JSON.stringify(state));
      }}>CLOSE</Button>
    </>;
```

File: examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+
+  return <>
+    <Paragraph>Hello Sidebar Component (AppId: {sdk.ids.app})</Paragraph>
+    <Paragraph>{`Time is ${state.timestamp}`}</Paragraph>
+    <Button onClick={() => {
+      dispatch({
+        type: 'SET_TIME',
+        payload: new Date().toString(),
+      });
+    }}>SET TIME</Button>
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx:32]:
```suggestion
      <Button
        onClick={() => {
          dispatch({
            type: 'SET_TIME',
            payload: new Date().toString(),
          });
        }}>
        SET TIME
      </Button>
```

File: examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+          serializedSidebarState: JSON.stringify(state),
+        },
+      });
+      if (serializedDialogState) {
+        dispatch({
+          type: 'SET_WHOLE_STATE',
+          payload: JSON.parse(serializedDialogState),
+        });
+      }
+    }}>OPEN DIALOG</Button>
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx:49]:
```suggestion
      <Button
        onClick={async () => {
          const serializedDialogState = await sdk.dialogs.openCurrentApp({
            shouldCloseOnEscapePress: false,
            shouldCloseOnOverlayClick: false,
            title: 'Dialog',
            minHeight: 360,
            parameters: {
              serializedSidebarState: JSON.stringify(state),
            },
          });
          if (serializedDialogState) {
            dispatch({
              type: 'SET_WHOLE_STATE',
              payload: JSON.parse(serializedDialogState),
            });
          }
        }}>
        OPEN DIALOG
      </Button>
```

File: examples/stateful-sidebar-with-dialog/src/model/reducerAndInitialSidebarState.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+
+
+export function reducer(state, { type, payload }) {
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/model/reducerAndInitialSidebarState.js:3]:
```suggestion
export function reducer(state, { type, payload }) {
```

File: examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+      });
+    }}>SET TIME</Button>
+    <Button
+      onClick={() => {
+        sdk.close(JSON.stringify(state));
+      }}>CLOSE</Button>
+    </>;
+};
+
+export default Dialog;
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx:39]:
```suggestion
import React, { useEffect, useReducer } from 'react';
import { Paragraph, Button } from '@contentful/f36-components';
import { /* useCMA, */ useSDK } from '@contentful/react-apps-toolkit';
import { reducer } from '../model/reducerAndInitialSidebarState';

const Dialog = () => {
  const sdk = useSDK();
  const { serializedSidebarState } = sdk.parameters.invocation || {};
  const [state, dispatch] = useReducer(reducer, {});
  /*
     To use the cma, inject it as follows.
     If it is not needed, you can remove the next line.
  */
  // const cma = useCMA();

  useEffect(() => {
    dispatch({
      type: 'SET_WHOLE_STATE',
      payload: JSON.parse(serializedSidebarState),
    });
  }, []);

  return (
    <>
      <Paragraph>Hello Dialog Component (AppId: {sdk.ids.app})</Paragraph>
      <Paragraph>{`Time is ${state.timestamp}`}</Paragraph>
      <Button
        onClick={() => {
          dispatch({
            type: 'SET_TIME',
            payload: new Date().toString(),
          });
        }}>
        SET TIME
      </Button>
      <Button
        onClick={() => {
          sdk.close(JSON.stringify(state));
        }}>
        CLOSE
      </Button>
    </>
  );
};

export default Dialog;

```

File: examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+              payload: JSON.parse(serializedDialogState),
+            });
+          }
+        }}>
+        OPEN DIALOG
+      </Button>
+  </>;
+};
+
+export default Sidebar;
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx:59]:
```suggestion
import React, { useEffect, useReducer } from 'react';
import { Paragraph, Button } from '@contentful/f36-components';
import { /* useCMA, */ useSDK } from '@contentful/react-apps-toolkit';
import { reducer, initialSidebarState } from '../model/reducerAndInitialSidebarState';

const Sidebar = () => {
  const sdk = useSDK();
  const [state, dispatch] = useReducer(reducer, {}, initialSidebarState);
  /*
     To use the cma, inject it as follows.
     If it is not needed, you can remove the next line.
  */
  // const cma = useCMA();

  useEffect(() => {
    dispatch({
      type: 'SET_TIME',
      payload: new Date().toString(),
    });
  }, []);

  const someParameter = 'some parameter';

  return (
    <>
      <Paragraph>Hello Sidebar Component (AppId: {sdk.ids.app})</Paragraph>
      <Paragraph>{`Time is ${state.timestamp}`}</Paragraph>
      <Button
        onClick={() => {
          dispatch({
            type: 'SET_TIME',
            payload: new Date().toString(),
          });
        }}>
        SET TIME
      </Button>
      <Button
        onClick={async () => {
          const serializedDialogState = await sdk.dialogs.openCurrentApp({
            shouldCloseOnEscapePress: false,
            shouldCloseOnOverlayClick: false,
            title: 'Dialog',
            minHeight: 360,
            parameters: {
              serializedSidebarState: JSON.stringify(state),
            },
          });
          if (serializedDialogState) {
            dispatch({
              type: 'SET_WHOLE_STATE',
              payload: JSON.parse(serializedDialogState),
            });
          }
        }}>
        OPEN DIALOG
      </Button>
    </>
  );
};

export default Sidebar;

```

File: examples/stateful-sidebar-with-dialog/src/model/reducerAndInitialSidebarState.js [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+  }
+
+  return state;
+}
+
+export function initialSidebarState() {
+  return {
+    timestamp: 'SET ME!'
+  };
+}
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/model/reducerAndInitialSidebarState.js:19]:
```suggestion
export function reducer(state, { type, payload }) {
  if (type === 'SET_TIME') {
    return {
      ...state,
      timestamp: payload,
    };
  }
  if (type === 'SET_WHOLE_STATE') {
    return payload;
  }

  return state;
}

export function initialSidebarState() {
  return {
    timestamp: 'SET ME!',
  };
}

```

File: examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+          sdk.close(JSON.stringify(state));
+        }}>
+        CLOSE
+      </Button>
+    </>
+  );
+};
+
+export default Dialog;
+
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx:47]:
```suggestion

export default Dialog;
```

**Comment by david-shibley-contentful** [comment]
I put together another PR with the exact same code and all the tests are passing: https://github.com/contentful/apps/pull/8831

File: examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx (hunk 9 above)
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx:47]:
```suggestion
  );
};

export default Dialog
```

File: examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+          }
+        }}>
+        OPEN DIALOG
+      </Button>
+    </>
+  );
+};
+
+export default Sidebar;
+
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx:62]:
```suggestion
};

export default Sidebar
```

File: examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+        onClick={() => {
+          sdk.close(JSON.stringify(state));
+        }}>
+        CLOSE
+      </Button>
+    </>
+  );
+};
+
+export default Dialog
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Dialog.jsx]:
```suggestion
export default Dialog;
```

File: examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+            });
+          }
+        }}>
+        OPEN DIALOG
+      </Button>
+    </>
+  );
+};
+
+export default Sidebar
```
Comment (david-shibley-contentful) [inline examples/stateful-sidebar-with-dialog/src/locations/Sidebar.jsx:61]:
```suggestion
export default Sidebar;
```

**Review by david-shibley-contentful — APPROVED** [review]

**Review by sarahlessner — APPROVED** [review]

**Comment by david-shibley-contentful** [comment]
Merged. Thanks for the contribution @Muzietto!

**Pull request merged**